server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -226,22 +238,45 @@ class AsyncAction {
         private final NodesRequest request;
         private final ActionListener<NodesResponse> listener;
         private final AtomicReferenceArray<Object> responses;
+        private final DiscoveryNode[] concreteNodes;
         private final AtomicInteger counter = new AtomicInteger();
         private final Task task;
 
         AsyncAction(Task task, NodesRequest request, ActionListener<NodesResponse> listener) {
             this.task = task;
             this.request = request;
             this.listener = listener;
-            if (request.concreteNodes() == null) {
+
+            // Check if concrete nodes are already available
+            if (request.concreteNodes() != null) {
+                this.responses = new AtomicReferenceArray<>(request.concreteNodes().length);
+
+                if (request.populateDiscoveryNodesInTransportRequest()) {
+                    this.concreteNodes = null;
```

Comment:
why this has to be set to null explicitly, it not initialized before this point.

Reply:
Since I have set the `concreteNodes` as final, it has to be initialised by the constructor. 
I have set it to `final` since we want the concrete nodes to be immutable at all cost since it decides where to route the requests to.